src/hashset.h

```diff
@@ -0,0 +1,205 @@
+ * contain a key. It can represent a key-value entry, or it can be just a key,
+ * if set semantics are desired.
+ *
+ * Terminology:
+ *
+ * hashset
+ *         An instance of the data structure.
+ *
+ * key
+ *         A key used for looking up an element in the hashset.
+ *
+ * element
+ *         An element in the hashset. This may be of the same type as the key,
+ *         or a struct containing a key and other fields.
+ *
+ * type
+ *         A struct containing callbacks, such as hash function, key comparison
+ *         function and how to get the key in an element.
+ */
+
+#include "fmacros.h"
+#include <stddef.h>
+#include <stdint.h>
+
+/* --- Opaque types --- */
+
+typedef struct hashset hashset;
+
+/* --- Non-opaque types --- */
+
+/* The hashsetType is a set of callbacks for a hashset. All callbacks are
+ * optional. With all callbacks omitted, the hashset is effectively a set of
+ * pointer-sized integers. */
+typedef struct {
+    /* If the type of an element is not the same as the type of a key used for
+     * lookup, this callback needs to return the key within an element. */
+    const void *(*elementGetKey)(const void *element);
+    /* Hash function. Defaults to hashing the bits in the pointer, effectively
+     * treating the pointer as an integer. */
+    uint64_t (*hashFunction)(const void *key);
+    /* Compare function, returns 0 if the keys are equal. Defaults to just
+     * comparing the pointers for equality. */
+    int (*keyCompare)(hashset *t, const void *key1, const void *key2);
+    /* Callback to free an element when it's overwritten or deleted.
+     * Optional. */
+    void (*elementDestructor)(hashset *t, void *elem);
+    /* Optional callback to control when resizing should be allowed. */
+    int (*resizeAllowed)(size_t moreMem, double usedRatio);
+    /* Invoked at the start of rehashing. Both tables are already created. */
+    void (*rehashingStarted)(hashset *t);
+    /* Invoked at the end of rehashing. Both tables still exist and are cleaned
+     * up after this callback. */
+    void (*rehashingCompleted)(hashset *t);
+    /* Allow a hashset to carry extra caller-defined metadata. The extra memory
+     * is initialized to 0. */
+    size_t (*getMetadataSize)(void);
+    /* Flag to disable incremental rehashing */
+    unsigned instant_rehashing : 1;
+    /* Allow the caller to store some data here in the type. It's useful for the
+     * rehashingStarted and rehashingCompleted callbacks. */
+    void *userdata;
+} hashsetType;
+
+typedef enum {
+    HASHSET_RESIZE_ALLOW = 0,
+    HASHSET_RESIZE_AVOID,
+    HASHSET_RESIZE_FORBID,
+} hashsetResizePolicy;
+
+typedef void (*hashsetScanFunction)(void *privdata, void *element);
+
+/* Constants */
+#define HASHSET_BUCKET_SIZE 64 /* bytes */
+
+/* Scan flags */
+#define HASHSET_SCAN_EMIT_REF (1 << 0)
+#define HASHSET_SCAN_SINGLE_STEP (1 << 2)
+
+typedef struct {
+    hashset *hashset;
+    long index;
+    int table;
+    int posInBucket;
```

Comment:
```suggestion
    int pos_in_bucket;
```
The most common convention we've used for struct field names is snake_case, although camelCase is also fairly common. I think we should adopt it officially. Maybe I'll go open a PR in the contributor file for it.

Reply:
If I change this one, I should change all...

What about the callbacks in hashsetType? They're all camelCase.

In dict, all struct fields are camelCase. I did the same here. I wanted the API to look as similar as possible, especially for hashsetType, but of course it can be changed too.

Reply:
I agree it should be all or nothing. I'll bring it up on monday. I think it's fine leaving it for now, I would just like to start moving more towards something that is standard.

Reply:
Changed all to snake case.